src/tree/updater_gpu_hist.cu

```diff
@@ -1326,23 +1305,50 @@ inline void DeviceShard<GradientSumT>::InitCompressedData(
 
 template <typename GradientSumT>
 inline void DeviceShard<GradientSumT>::CreateHistIndices(
-    const SparsePage& row_batch, size_t row_stride, int null_gidx_value) {
+    const SparsePage &row_batch,
+    const common::HistCutMatrix &hmat,
+    const std::vector<size_t> &shard_allocations,
+    int rows_per_batch) {
+  // Has any been allocated for me in this batch?
+  size_t num_elems = shard_allocations[shard_idx];
```

Comment:
Nitpick: use vector.at() outside of hot loops for better error messages.

Reply:
are you suggesting because it has the potential to throw an exception?